src/osp/Resource/Package.h

```diff
@@ -90,121 +71,143 @@ StrViewPair_t decompose_str(std::string_view path, const char delim);
-//    Resource<T>* get_resource(unsigned resIndex);
-
+    /**
+     * Initialize and add a resource to store in this package.
+     *
+     * @tparam TYPE_T   Type of resource to store
+     *
+     * @param path [in] String used to identify this resource
+     * @param args [in] Arguments to pass to TYPE_T constructor
+     * @return Reference counted dependency to new resource
+     */
     template<class TYPE_T, typename ... ARGS_T>
     DependRes<TYPE_T> add(std::string_view path, ARGS_T&& ... args);
 
+    /**
+     * Get a resource by path identifier
+     *
+     * @tparam TYPE_T   Type of resource to get
+     *
+     * @param path [in] String used to identify the target resource
+     * @return Reference counted dependency to the found resource; empty if not
+     *         found.
+     */
     template<class TYPE_T>
-    DependRes<TYPE_T> get(std::string_view path);
+    DependRes<TYPE_T> get(std::string_view path) noexcept;
+
+    // TODO: function for removing specific resources
 
+    /**
+     * Remove all stored resource of the specified type
+     *
+     * @tparam TYPE_T Type of resource to clear
+     */
     template<class TYPE_T>
-    void clear();
+    void clear() noexcept;
 
+    // Stores maps of resoures of a specified type
+    template<class TYPE_T>
     struct GroupType
     {
-        virtual ~GroupType() = default;
-    };
+        GroupType() noexcept = default;
+        GroupType(GroupType&& move) noexcept = default;
+        GroupType& operator=(GroupType&& move) noexcept = default;
+
+        // Needs to be copyable to store in entt::any, but copy is not allowed
+        // entt experimental branch allows move-only types.
+        GroupType(GroupType const& copy) { assert(0); };
 
-    template<class TYPE_T>
-    struct GroupTypeRes : public GroupType
-    {
-        ~GroupTypeRes() = default;
-        //std::vector<TYPE_T> m_types;
         std::map<std::string, Resource<TYPE_T>, std::less<>> m_resources;
     };
 
-    ResPrefix_t get_prefix() const { return m_prefix; }
+    ResPrefix_t get_prefix() { return m_prefix; }
 
 private:
 
-    std::map<entt::id_type, std::unique_ptr<GroupType>> m_groups;
+    std::vector<entt::any> m_groups;
 
     std::string m_packageName;
 
     ResPrefix_t m_prefix;
 
     std::string m_displayName;
-
 };
 
 template<class TYPE_T, typename ... ARGS_T>
 DependRes<TYPE_T> Package::add(std::string_view path,
                                ARGS_T&& ... args)
 {
-    // this should create a blank if it doesn't exist yet
-    //std::any& groupAny(m_groups[entt::type_info<TYPE_T>::id()]);
-    std::unique_ptr<GroupType>& groupAny(m_groups[entt::type_info<TYPE_T>::id()]);
+    // Runtime generated (global) sequential ID
+    // First resource added is 0, next is 1, then 2, etc...
+    const uint32_t resId = resource_id::type<TYPE_T>;
 
+    // Resize m_groups to ensure that resId a valid index
+    if (m_groups.size() <= resId)
+    {
+        m_groups.resize(resId + 1);
```

Comment:
Not a big fan of this. You're causing a new alloction each time. No doubling algorithm for the capacity. 

Use push_back instead? Or at least a call to reserve before resize (which does the doubling algom)

Reply:
These IDs are generated per-type. This is called very few times when types are first added to the Package. EnTT does a similar thing with its pools on the inside. Maybe I should change resId to resTypeId, and just make this more clear in general.

Reply:
My point was about the reallocations of the vector, since you're circumventing the allocators doubling heuristic.

No way to use push_back for this?

Reply:
Nope, resId can be any number above the current capacity of the m_groups vector. It's possible to resize to a slightly higher capacity, but this is a minor optimization anyways.